### PR TITLE
FOG-275 Fog Signature Scheme

### PR DESCRIPTION
**Pull request by jcape**
Soundtrack of this PR: [link to song that really fits the mood of this PR](https://www.youtube.com/watch?v=MJuyn0WAYNI)

### Motivation

We need to be able to verify fog signatures.

### In this PR

This PR implements the fog signature scheme and ties it into verification. The layering and dependencies are [somewhat complex](https://bit.ly/2Ng38dC), but basically there are two "base" crates, `mc-fog-sig-authority` and `mc-fog-sig-report` crates, which implement the authority and report signature/validation schemes in a no-std way in the (current) Ristretto/Ed algorithms.

Then there's `mc-account-keys`, which depends on `mc-fog-sig-authority` to produce it's signatures, and implements verification for `PublicAddress` (this is done here because the signature currently happens inside the subaddress generation method on `AccountKey`). On top of `mc-account-keys` and the `mc-fog-sig-report` crate, there's `mc-fog-sig`, which provides a top-to-bottom verifier. This verifies the authority, chain, and report signatures all in one shot. This is *not* no-std friendly, because it includes PEM and X509 handling, which are not no-std. This is facilitated by the digestible signatures and x509 handling crates, the latter of which has been rewritten to use `x509-signature` instead of `x509-parser`, because the latter had no reasonable way of accessing the raw `subjectPublicKeyInfo` bytes.

This is used by `mc-fog-report-validation` to verify the chain of trust from recipient's `PublicAddress` to the report data structures.

Additionally, the Fog Report API (and corresponding entries in `mc-fog-types`) have been extended to add support for a signature and chain (as a list of DER-encoded `bytes`), and the verification report uses a proper protobuf type, rather than a serde-encoded `bytes`. `VerificationReport` has been added to `external.proto`.

### Future Work
* More unit testing, particularly round-trips through 
* Updating the report server to use this to produce these signatures (internal will be broken until that is complete).

### CODE REVIEW

File: api/src/convert/mod.rs [hunk 1]
```diff
@@ -35,34 +35,10 @@ mod tx_out_confirmation_number;
 mod tx_out_membership_element;
 mod tx_out_membership_proof;
 mod tx_prefix;
+mod verification_report;
 mod watcher;
 
-pub use account_key::*;
```
Comment (eranrund) [inline api/src/convert/mod.rs:40]: Curious why those were removed. Do the `impl From/TryFrom` work without the use statements?

File: api/src/convert/verification_report.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    fn from(src: &external::VerificationReport) -> Self {
+        VerificationReport {
+            sig: VerificationSignature::from(src.get_sig()),
+            chain: src.get_chain().to_vec(),
+            http_body: src.get_http_body().to_owned(),
+        }
+    }
+}
+
+// #[cfg(test)]
```
Comment (eranrund) [inline api/src/convert/verification_report.rs:28]: Missing tests?

File: crypto/x509/test-vectors/generate.sh [hunk 3]
```diff
@@ -308,8 +308,8 @@ cat "${OK_PENULTIMATE_DIR}/private/leaf.key" > "${OUTPUT_BASE_DIR}/fail_missing_
 	-config "$OPENSSL_CNF" \
 	-name ok_penultimate \
 	-extensions leaf_cert \
-	-startdate 010101000000Z \
-	-enddate 100101000001Z \
+	-startdate 20010101000000Z \
```
Comment (eranrund) [inline crypto/x509/test-vectors/generate.sh:311]: Why was this needed?

File: crypto/x509/test-vectors/src/lib.rs [hunk 4]
```diff
@@ -26,11 +25,13 @@ fn get_leaf_key(name: &str) -> Ed25519Pair {
     path.set_extension("key");
 
     let pem_string = fs::read_to_string(path).expect("Could not read certificate chain");
-    let (_, pem_data) =
-        pem::parse_x509_pem(pem_string.as_ref()).expect("Could not parse PEM string");
-
+    eprintln!("Parsing PEM key");
```
Comment (eranrund) [inline crypto/x509/test-vectors/src/lib.rs:28]: Leftovers? Or useful to have around?

File: crypto/x509/utils/src/lib.rs [hunk 5]
```diff
@@ -3,65 +3,24 @@
...
-
-/// Anything which can be referenced as a byte slice gets into_iter_pem()
-impl<T: AsRef<[u8]>> PemStringIterable for T {
-    fn into_pem_iter(self) -> PemStringIter<T> {
-        PemStringIter::new(self)
-    }
-}
+use x509_signature::{ASN1Time, Error as X509Error, X509Certificate};
 
 /// An iterator of [`X509Certificate`] objects over a slice of [`Pem`] objects.
```
Comment (eranrund) [inline crypto/x509/utils/src/lib.rs:14]: Is this comment still correct? Looks like its no longer `Pem` objects

File: crypto/x509/utils/src/lib.rs [hunk 6]
```diff
@@ -136,32 +100,32 @@ impl<'a, T: AsRef<[X509Certificate<'a>]>> X509CertificateChain for T {
...
+            if let Some(prev_cert) = previous {
+                cert.check_issued_by(prev_cert)?;
+            } else {
+                cert.check_self_issued()?;
+            }
+
+            let timestamp = SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)?
+                .as_secs() as i64;
+            eprintln!(
```
Comment (eranrund) [inline crypto/x509/utils/src/lib.rs:112]: leftover?

File: fog/report/validation/src/ingest_report.rs [hunk 7]
```diff
@@ -24,23 +21,10 @@ impl IngestReportVerifier {
...
-
-        let mut verifier = self.verifier.clone();
-        verifier
-            .sign_type(QuoteSignType::Linkable)
-            .report_data(&masked_report_data);
-
-        verifier.verify(&remote_report)?;
-
-        Ok(report_pubkey)
+        Ok(RistrettoPublic::try_from(&report_data_bytes[32..64])?)
```
Comment (eranrund) [inline fog/report/validation/src/ingest_report.rs:27]: Is there a guarantee there are 64 bytes in the report data?

File: fog/sig/authority/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+mod ristretto;
+
+use core::fmt::{Debug, Display};
+use signature::Signature;
+
+/// Retrieve the canonical signing context byte string.
+///
+/// This is intended to be used by crate-remote implementations of the
+/// signature who want a "standard"
+pub fn context() -> &'static [u8] {
```
Comment (eranrund) [inline fog/sig/authority/src/lib.rs:22]: Wondering why is this not a `const`?

File: fog/sig/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+//!  1. Allow a recipient to sign a Fog Authority (subjectPublicKeyInfo)
+//!  1. Allow the fog report server to sign a list of responses.
+//!  1. Allow a sender to verify and extract the ingest key from a user's
+//!     PublicAddress and a ReportResponse object(s).
+//!
+//! # Fog Authority Signatures
+//!
+//! Fog authority signatures are created by the owner of an account, in order to
+//! delegate transaction monitoring to an enclave using the on-chain fog
+//! accounts hints. In order to accomplish this, the user signs the  DER-encoded
```
Comment (eranrund) [inline fog/sig/src/lib.rs:14]:
```suggestion
//! accounts hints. In order to accomplish this, the user signs the DER-encoded
```

File: fog/sig/src/public_address.rs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        )
+        .collect::<Vec<X509Certificate>>();
+
+        // Verify the authority signature
+        let authority_sig = self
+            .fog_authority_sig()
+            .ok_or(Error::NoSignature)?
+            .try_into()?;
+
+        self.verify_authority(&certs[0].subject_public_key_info().spki(), &authority_sig)
```
Comment (eranrund) [inline fog/sig/src/public_address.rs:43]: If the report contained a bunch of invalid signatures (or just the first one was invalid) then `certs` would be empty and we'll panic.

**Review by eranrund — APPROVED** [review]
Nice! I got some small questions/nits/minor concerns but nothing that should get in the way of getting this merged once addressed.

File: account-keys/src/account_keys.rs [hunk 11]
```diff
@@ -171,6 +175,15 @@ impl PublicAddress {
     }
 }
 
+impl AuthorityVerifier for PublicAddress {
+    type Sig = <RistrettoPublic as AuthorityVerifier>::Sig;
+    type Error = <RistrettoPublic as AuthorityVerifier>::Error;
+
+    fn verify_authority(&self, spki_bytes: &[u8], sig: &Self::Sig) -> Result<(), Self::Error> {
```
Comment (garbageslam) [inline account-keys/src/account_keys.rs:182]: why does this take a sig as a parameter? shouldn't it take the sig from Self ?

File: fog/sig/authority/src/ristretto.rs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+use alloc::{format, string::String};
+use mc_crypto_keys::{RistrettoPrivate, RistrettoPublic, RistrettoSignature};
+
+impl Signer for RistrettoPrivate {
+    type Sig = RistrettoSignature;
+    type Error = String;
+
+    fn sign_authority(&self, spki_bytes: &[u8]) -> Result<Self::Sig, String> {
+        Ok(self.sign_schnorrkel(super::context(), spki_bytes))
+    }
```
Comment (garbageslam) [inline fog/sig/authority/src/ristretto.rs:19]: Does this function need to return result? it looks like it is infallible, and it looks like the caller is unwrapping the error right now

File: fog/report/validation/src/lib.rs [hunk 13]
```diff
@@ -100,18 +98,17 @@ impl FogPubkeyResolver for FogResolver {
...
                     if report_id == report.fog_report_id {
-                        // TODO validate x509 chain and recipient.fog_authority_sig here,
-                        // However, probably skip that if uri scheme is "insecure-fog"?
-                        // TODO this should not use mc_util_serial::deserialize, we should use prost
-                        let remote_report: VerificationReport =
-                            mc_util_serial::deserialize(&report.report)?;
-                        let pubkey = self.verifier.validate_ingest_ias_report(remote_report)?;
+                        let pubkey = self
+                            .verifier
+                            .validate_ingest_ias_report(report.report.clone())?;
```
Comment (jcape) [inline fog/report/validation/src/lib.rs:111]: Possible issue

File: api/src/convert/mod.rs (hunk 1 above)
Comment (jcape) [reply api/src/convert/mod.rs:40]: Yep

**Review by garbageslam — APPROVED** [review]

**Review by sugargoat — APPROVED** [review]

File: crypto/x509/test-vectors/generate.sh (hunk 3 above)
Comment (jcape) [reply crypto/x509/test-vectors/generate.sh:311]: ASN1Time is the devil ;-)

File: fog/report/validation/src/ingest_report.rs (hunk 7 above)
Comment (jcape) [reply fog/report/validation/src/ingest_report.rs:27]: Yeah, it's a fixed length in the SGX SDK, and they do a bunch of "cast these bytes to this structure"-style C, so if they did change this it would be a severe breaking change of the SGX SDK.

File: account-keys/src/account_keys.rs (hunk 11 above)
Comment (jcape) [reply account-keys/src/account_keys.rs:182]: Under all the covers, the trait is implemented primarily on the `RistrettoPublic` type, so the trait needs to have the sig passed in.

File: fog/sig/authority/src/ristretto.rs (hunk 12 above)
Comment (jcape) [reply fog/sig/authority/src/ristretto.rs:19]: The `signature::Signer::sign()` crate allows fallible signatures, because it's designed to work with HSMs (where you can fail to contact the HSM, because, e.g. the thing isn't plugged in). I'm carrying that through for all this signature stuff to allow for more secure key storage options in the future.
